src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+            }
+          } // for i
+        } // for j
+        result(x, y) = (sum * psf_total) / psf_portion;
+      } // for x
+    } // for y
+
+    // avoid copying the results back
+    image = std::move(result);
+  }
+
+
+#ifdef HAVE_CUDA
+  // Performs convolution between an image represented as an array of floats
+  // and a PSF on a GPU device.
+  extern "C" void deviceConvolve(float* source_img, float* result_img,
+                                 int width, int height, float* psf_kernel,
+                                 int psf_size, int psf_dim, int psf_radius,
+                                 float psf_sum);
+#endif
+
+
+  void RawImageEigen::convolve(PSF psf) {
+#ifdef HAVE_CUDA
+    deviceConvolve(image.data(), image.data(),
+                   get_width(), get_height(), psf.data(),
+                   psf.get_size(), psf.get_dim(), psf.get_radius(), psf.get_sum());
+#else
+    convolve_cpu(psf);
+#endif
+  }
+
+
+  void RawImageEigen::apply_bitmask(const ImageI& bitmask) {
+    image = (bitmask.array() == 1).select(NO_DATA, image);
+  }
+
+
+  void RawImageEigen::apply_mask1(const ImageI& mask, const int flags) {
+    // I'm not sure a copy is avoidable here, we can pass a non-const Ref
+    // and then edit the mask in-place, but that would edit the Python-side
+    // numpy array as well....
+    auto bitmask = mask.unaryExpr( [&flags](int x){return flags & x;} );
+    apply_bitmask(bitmask);
+  }
+
+
+  // Imma be honest, this function makes no sense to me... I think I transcribed
+  // it right, but I'm not sure? If a flag is in mask, it's masked unless it's in
+  // exceptions? But why do we have two lists? Is the example above better than this?
+  void RawImageEigen::apply_mask(int flags, const std::vector<int>& exceptions,
+                                 const RawImageEigen& mask) {
+    const int npixels = get_npixels();
+    assert(npixels == mask.get_npixels());
+
+    // evaluated as lvalue because of auto (hopefully)
+    auto mask_unravelled = mask.image.reshaped();
+    auto image_unravelled = image.reshaped();
+
+    for (unsigned int p = 0; p < npixels; ++p) {
+      int pix_flags = static_cast<int>(mask_unravelled(p));
+      bool is_exception = false;
+      for (auto& e : exceptions) is_exception = is_exception || e == pix_flags;
+      if (!is_exception && ((flags & pix_flags) != 0)) image_unravelled(p) = NO_DATA;
+    }
+  }
+
+
+  /* This implementation of grow_mask is optimized for steps > 1
+     (which is how the code is generally used. If you are only
+     growing the mask by 1, the extra copy will be a little slower.
+  */
+  void RawImageEigen::grow_mask(const unsigned steps) {
+    const int num_array = width * height;
+
+    // create a bitmask out of the image array
+    ImageI bitmask = ImageI::Zero(height, width);
+    bitmask = (image.array() == NO_DATA).select(1, bitmask);
+
+    // convolution with a dilating operator is probably the
```

Comment:
I forgot during out discussion, but convolution won't work here if we stay consistent with the original grow mask functionality.

Reply:
Postponed for some other PR, reverted to old behavior, same as above.